### PR TITLE
feat(Columns): Add configurable widths, gutters & responsive layout

### PR DESCRIPTION
**Pull request by michaeltaranto**
Loading up `Columns` with a bunch of features to make it usable for everything from form layout to page layout.

#### `Columns`
- `collapse={boolean}` — Collapse on mobile or preserve columns
- `reverse={boolean}` — Reverse order of columns
- `gutter={SpaceX}` — Configurable gutter widths, accepts all the same horizontal props as `Box`, e.g. `small`, `large`, etc
#### `Column`
- `width` — Supports fitting to content and fraction of screen sizes down to fifths, e.g. `width="1/5"`

## Previews
- [Columns docs](https://braid-design-system--1fd2727ba3ef195cd9ea04bc1e8235ea9d98cd70.surge.sh/components/Columns)
- [Column docs](https://braid-design-system--1fd2727ba3ef195cd9ea04bc1e8235ea9d98cd70.surge.sh/components/Column)
- [Playroom](https://braid-design-system--1fd2727ba3ef195cd9ea04bc1e8235ea9d98cd70.surge.sh/playroom/#?code=PEJveCBiYWNrZ3JvdW5kPSJzZWNvbmRhcnlMaWdodCIgc3R5bGU9e3sgaGVpZ2h0OiAiMTAwdmgiIH19PgogIDxDYXJkPgogICAgPEhlYWRpbmcgbGV2ZWw9IjIiPlN0YW5kYXJkPC9IZWFkaW5nPgogICAgPERpdmlkZXIgLz4KICAgIDxiciAvPgogICAgPEhlYWRpbmcgbGV2ZWw9IjQiPkRlZmF1bHQ8L0hlYWRpbmc-CiAgICA8Q29sdW1ucz4KICAgICAgPENvbHVtbj4KICAgICAgICA8Qm94IGJhY2tncm91bmQ9InNlbGVjdGlvbiIgcGFkZGluZz0ic21hbGwiPgogICAgICAgICAgPFRleHQgYmFzZWxpbmU9e2ZhbHNlfT5BPC9UZXh0PgogICAgICAgIDwvQm94PgogICAgICA8L0NvbHVtbj4KICAgICAgPENvbHVtbj4KICAgICAgICA8Qm94IGJhY2tncm91bmQ9InNlbGVjdGlvbiIgcGFkZGluZz0ic21hbGwiPgogICAgICAgICAgPFRleHQgYmFzZWxpbmU9e2ZhbHNlfT5CPC9UZXh0PgogICAgICAgIDwvQm94PgogICAgICA8L0NvbHVtbj4KICAgIDwvQ29sdW1ucz4KICAgIDxiciAvPgogICAgPEhlYWRpbmcgbGV2ZWw9IjQiPlJldmVyc2U8L0hlYWRpbmc-CiAgICA8Q29sdW1ucyByZXZlcnNlPgogICAgICA8Q29sdW1uPgogICAgICAgIDxCb3ggYmFja2dyb3VuZD0ic2VsZWN0aW9uIiBwYWRkaW5nPSJzbWFsbCI-CiAgICAgICAgICA8VGV4dCBiYXNlbGluZT17ZmFsc2V9PkE8L1RleHQ-CiAgICAgICAgPC9Cb3g-CiAgICAgIDwvQ29sdW1uPgogICAgICA8Q29sdW1uPgogICAgICAgIDxCb3ggYmFja2dyb3VuZD0ic2VsZWN0aW9uIiBwYWRkaW5nPSJzbWFsbCI-CiAgICAgICAgICA8VGV4dCBiYXNlbGluZT17ZmFsc2V9PkI8L1RleHQ-CiAgICAgICAgPC9Cb3g-CiAgICAgIDwvQ29sdW1uPgogICAgPC9Db2x1bW5zPgogIDwvQ2FyZD4KICA8Q2FyZD4KICAgIDxIZWFkaW5nIGxldmVsPSIyIj5Db2xsYXBzZTwvSGVhZGluZz4KICAgIDxEaXZpZGVyIC8-CiAgICA8YnIgLz4KICAgIDxIZWFkaW5nIGxldmVsPSI0Ij5EZWZhdWx0PC9IZWFkaW5nPgogICAgPENvbHVtbnMgY29sbGFwc2U-CiAgICAgIDxDb2x1bW4-CiAgICAgICAgPEJveCBiYWNrZ3JvdW5kPSJzZWxlY3Rpb24iIHBhZGRpbmc9InNtYWxsIj4KICAgICAgICAgIDxUZXh0IGJhc2VsaW5lPXtmYWxzZX0-QTwvVGV4dD4KICAgICAgICA8L0JveD4KICAgICAgPC9Db2x1bW4-CiAgICAgIDxDb2x1bW4-CiAgICAgICAgPEJveCBiYWNrZ3JvdW5kPSJzZWxlY3Rpb24iIHBhZGRpbmc9InNtYWxsIj4KICAgICAgICAgIDxUZXh0IGJhc2VsaW5lPXtmYWxzZX0-QjwvVGV4dD4KICAgICAgICA8L0JveD4KICAgICAgPC9Db2x1bW4-CiAgICA8L0NvbHVtbnM-CiAgICA8YnIgLz4KICAgIDxIZWFkaW5nIGxldmVsPSI0Ij5SZXZlcnNlPC9IZWFkaW5nPgogICAgPENvbHVtbnMgY29sbGFwc2UgcmV2ZXJzZT4KICAgICAgPENvbHVtbj4KICAgICAgICA8Qm94IGJhY2tncm91bmQ9InNlbGVjdGlvbiIgcGFkZGluZz0ic21hbGwiPgogICAgICAgICAgPFRleHQgYmFzZWxpbmU9e2ZhbHNlfT5BPC9UZXh0PgogICAgICAgIDwvQm94PgogICAgICA8L0NvbHVtbj4KICAgICAgPENvbHVtbj4KICAgICAgICA8Qm94IGJhY2tncm91bmQ9InNlbGVjdGlvbiIgcGFkZGluZz0ic21hbGwiPgogICAgICAgICAgPFRleHQgYmFzZWxpbmU9e2ZhbHNlfT5CPC9UZXh0PgogICAgICAgIDwvQm94PgogICAgICA8L0NvbHVtbj4KICAgIDwvQ29sdW1ucz4KICA8L0NhcmQ-CiAgPENhcmQ-CiAgICA8SGVhZGluZyBsZXZlbD0iMiI-d2lkdGg8L0hlYWRpbmc-CiAgICA8RGl2aWRlciAvPgogICAgPGJyIC8-CiAgICA8SGVhZGluZyBsZXZlbD0iNCI-RGVmYXVsdDwvSGVhZGluZz4KICAgIDxDb2x1bW5zPgogICAgICA8Q29sdW1uPgogICAgICAgIDxCb3ggYmFja2dyb3VuZD0ic2VsZWN0aW9uIiBwYWRkaW5nPSJzbWFsbCI-CiAgICAgICAgICA8VGV4dCBiYXNlbGluZT17ZmFsc2V9PkE8L1RleHQ-CiAgICAgICAgPC9Cb3g-CiAgICAgIDwvQ29sdW1uPgogICAgICA8Q29sdW1uPgogICAgICAgIDxCb3ggYmFja2dyb3VuZD0ic2VsZWN0aW9uIiBwYWRkaW5nPSJzbWFsbCI-CiAgICAgICAgICA8VGV4dCBiYXNlbGluZT17ZmFsc2V9PkI8L1RleHQ-CiAgICAgICAgPC9Cb3g-CiAgICAgIDwvQ29sdW1uPgogICAgPC9Db2x1bW5zPgogICAge1siMS8zIiwgIjIvMyIsICIxLzQiLCAiMi80IiwgIjMvNCIsICIxLzUiLCAiMi81IiwgIjMvNSIsICI0LzUiXS5tYXAoCiAgICAgIHdpZHRoID0-ICgKICAgICAgICA8PgogICAgICAgICAgPGJyIC8-CiAgICAgICAgICA8SGVhZGluZyBsZXZlbD0iNCI-e3dpZHRofTwvSGVhZGluZz4KICAgICAgICAgIDxDb2x1bW5zPgogICAgICAgICAgICA8Q29sdW1uIHdpZHRoPXt3aWR0aH0-CiAgICAgICAgICAgICAgPEJveCBiYWNrZ3JvdW5kPSJzZWxlY3Rpb24iIHBhZGRpbmc9InNtYWxsIj4KICAgICAgICAgICAgICAgIDxUZXh0IGJhc2VsaW5lPXtmYWxzZX0-QTwvVGV4dD4KICAgICAgICAgICAgICA8L0JveD4KICAgICAgICAgICAgPC9Db2x1bW4-CiAgICAgICAgICAgIDxDb2x1bW4-CiAgICAgICAgICAgICAgPEJveCBiYWNrZ3JvdW5kPSJzZWxlY3Rpb24iIHBhZGRpbmc9InNtYWxsIj4KICAgICAgICAgICAgICAgIDxUZXh0IGJhc2VsaW5lPXtmYWxzZX0-QjwvVGV4dD4KICAgICAgICAgICAgICA8L0JveD4KICAgICAgICAgICAgPC9Db2x1bW4-CiAgICAgICAgICA8L0NvbHVtbnM-CiAgICAgICAgPC8-CiAgICAgICkKICAgICl9CiAgICA8YnIgLz4KICAgIDxIZWFkaW5nIGxldmVsPSI0Ij5Db250ZW50IHdpZHRoPC9IZWFkaW5nPgogICAgPENvbHVtbnM-CiAgICAgIDxDb2x1bW4gd2lkdGg9ImNvbnRlbnQiPgogICAgICAgIDxCb3ggYmFja2dyb3VuZD0ic2VsZWN0aW9uIiBwYWRkaW5nPSJzbWFsbCI-CiAgICAgICAgICA8VGV4dCBiYXNlbGluZT17ZmFsc2V9PkNvbnRlbnQ8L1RleHQ-CiAgICAgICAgPC9Cb3g-CiAgICAgIDwvQ29sdW1uPgogICAgICA8Q29sdW1uPgogICAgICAgIDxCb3ggYmFja2dyb3VuZD0ic2VsZWN0aW9uIiBwYWRkaW5nPSJzbWFsbCI-CiAgICAgICAgICA8VGV4dCBiYXNlbGluZT17ZmFsc2V9PkZsZXhpYmxlPC9UZXh0PgogICAgICAgIDwvQm94PgogICAgICA8L0NvbHVtbj4KICAgIDwvQ29sdW1ucz4KICA8L0NhcmQ-CiAgPENhcmQ-CiAgICA8SGVhZGluZyBsZXZlbD0iMiI-R3V0dGVyczwvSGVhZGluZz4KICAgIDxEaXZpZGVyIC8-CiAgICA8YnIgLz4KICAgIDxIZWFkaW5nIGxldmVsPSI0Ij5EZWZhdWx0PC9IZWFkaW5nPgogICAgPENvbHVtbnM-CiAgICAgIDxDb2x1bW4-CiAgICAgICAgPEJveCBiYWNrZ3JvdW5kPSJzZWxlY3Rpb24iIHBhZGRpbmc9InNtYWxsIj4KICAgICAgICAgIDxUZXh0IGJhc2VsaW5lPXtmYWxzZX0-QTwvVGV4dD4KICAgICAgICA8L0JveD4KICAgICAgPC9Db2x1bW4-CiAgICAgIDxDb2x1bW4-CiAgICAgICAgPEJveCBiYWNrZ3JvdW5kPSJzZWxlY3Rpb24iIHBhZGRpbmc9InNtYWxsIj4KICAgICAgICAgIDxUZXh0IGJhc2VsaW5lPXtmYWxzZX0-QjwvVGV4dD4KICAgICAgICA8L0JveD4KICAgICAgPC9Db2x1bW4-CiAgICA8L0NvbHVtbnM-CiAgICA8YnIgLz4KICAgIDxIZWFkaW5nIGxldmVsPSI0Ij5UaWdodDwvSGVhZGluZz4KICAgIDxDb2x1bW5zIGd1dHRlcj0ic21hbGwiPgogICAgICA8Q29sdW1uPgogICAgICAgIDxCb3ggYmFja2dyb3VuZD0ic2VsZWN0aW9uIiBwYWRkaW5nPSJzbWFsbCI-CiAgICAgICAgICA8VGV4dCBiYXNlbGluZT17ZmFsc2V9PkE8L1RleHQ-CiAgICAgICAgPC9Cb3g-CiAgICAgIDwvQ29sdW1uPgogICAgICA8Q29sdW1uPgogICAgICAgIDxCb3ggYmFja2dyb3VuZD0ic2VsZWN0aW9uIiBwYWRkaW5nPSJzbWFsbCI-CiAgICAgICAgICA8VGV4dCBiYXNlbGluZT17ZmFsc2V9PkI8L1RleHQ-CiAgICAgICAgPC9Cb3g-CiAgICAgIDwvQ29sdW1uPgogICAgPC9Db2x1bW5zPgogICAgPGJyIC8-CiAgICA8SGVhZGluZyBsZXZlbD0iNCI-Tm8gR3V0dGVyPC9IZWFkaW5nPgogICAgPENvbHVtbnMgZ3V0dGVyPSJub25lIj4KICAgICAgPENvbHVtbj4KICAgICAgICA8Qm94IGJhY2tncm91bmQ9InNlbGVjdGlvbiIgcGFkZGluZz0ic21hbGwiPgogICAgICAgICAgPFRleHQgYmFzZWxpbmU9e2ZhbHNlfT5BPC9UZXh0PgogICAgICAgIDwvQm94PgogICAgICA8L0NvbHVtbj4KICAgICAgPENvbHVtbj4KICAgICAgICA8Qm94IGJhY2tncm91bmQ9InNlbGVjdGlvbiIgcGFkZGluZz0ic21hbGwiPgogICAgICAgICAgPFRleHQgYmFzZWxpbmU9e2ZhbHNlfT5CPC9UZXh0PgogICAgICAgIDwvQm94PgogICAgICA8L0NvbHVtbj4KICAgIDwvQ29sdW1ucz4KICAgIDxiciAvPgogICAgPEhlYWRpbmcgbGV2ZWw9IjQiPlJldmVyc2VkIEd1dHRlcjwvSGVhZGluZz4KICAgIDxDb2x1bW5zIHJldmVyc2U-CiAgICAgIDxDb2x1bW4-CiAgICAgICAgPEJveCBiYWNrZ3JvdW5kPSJzZWxlY3Rpb24iIHBhZGRpbmc9InNtYWxsIj4KICAgICAgICAgIDxUZXh0IGJhc2VsaW5lPXtmYWxzZX0-QTwvVGV4dD4KICAgICAgICA8L0JveD4KICAgICAgPC9Db2x1bW4-CiAgICAgIDxDb2x1bW4-CiAgICAgICAgPEJveCBiYWNrZ3JvdW5kPSJzZWxlY3Rpb24iIHBhZGRpbmc9InNtYWxsIj4KICAgICAgICAgIDxUZXh0IGJhc2VsaW5lPXtmYWxzZX0-QjwvVGV4dD4KICAgICAgICA8L0JveD4KICAgICAgPC9Db2x1bW4-CiAgICA8L0NvbHVtbnM-CiAgPC9DYXJkPgo8L0JveD4K)

### CODE REVIEW

File: generate-component-docs/generate.ts [hunk 1]
```diff
@@ -12,11 +12,16 @@ const componentsFile = path.join(__dirname, '../lib/components/index.ts');
...
+const columnType =
+  'ReactElement<ColumnProps, string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)> | null) | (new (props: any) => Component<any, any, any>)>';
+
 const stringAliases: Record<string, string> = {
   [reactNodeType]: 'ReactNode',
   // with an explicit alias 'boolean' becomes a union of 'true' | 'false'
   boolean: 'boolean',
   CSSProperties: 'CSSProperties',
+  [columnType]: 'Column',
+  [`${columnType}[]`]: 'Column[]',
```
Comment (markdalgleish) [inline generate-component-docs/generate.ts]: Nice! 👏

File: lib/components/Column/Column.treat.ts [hunk 2]
```diff
@@ -1,7 +1,24 @@
...
 
 export const column = style({
-  flexBasis: 0,
-  flexGrow: 1,
   minWidth: 0,
+  overflow: 'hidden',
 });
+
+export const fit = style({ flexShrink: 0 });
+const grow = style({ flexGrow: 1 });
```
Comment (mattsjones) [inline lib/components/Column/Column.treat.ts]: Feels like a good time to move this to `Box`?

Comment (patrickliu-sa) [reply lib/components/Column/Column.treat.ts]: > Feels like a good time to move this to `Box`?

It might not work if the parent is not a flexBox? 🤔

Comment (markdalgleish) [reply lib/components/Column/Column.treat.ts]: IMO that's okay. `Box` is conceptually at the same layer as CSS, and CSS has the same issue (if we call it that).

**Comment by tgdev** [comment]
This is awesome @michaeltaranto 2 questions as a consumer of this component...

re: new `reverse` prop, could it accept an array for breakpoints? eg: mobile keep column order, desktop reverse (or vice versa)

Is it possible to add a vertical alignment prop? The use case being Image in col1 content in col2 with content being centre aligned to middle of image (height of col1)

**Comment by markdalgleish** [comment]
@tgdev Regard the `reverse` prop, we wanted to make sure that document order was the same as mobile order, tab order and screen reader order. We only allow you to reverse the columns visually on large screens.

**Comment by markdalgleish** [comment]
@tgdev We're definitely keen on looking into alignment, but as a separate piece of work. It needs some more thinking in terms of how it fits into the current architecture. In the meantime, you're free to align content within columns manually as you typically would.

File: lib/components/Column/Column.docs.tsx [hunk 3]
```diff
@@ -14,38 +20,56 @@ const docs: ComponentDocs = {
...
-                paddingTop="small"
-                paddingBottom="small"
-                paddingLeft="small"
-              >
-                <Text baseline={false}>Column</Text>
-              </Box>
+              <Content>Fluid</Content>
+            </HideCode>
+          </Column>
+          <Column width="1/3">
```
Comment (markdalgleish) [inline lib/components/Column/Column.docs.tsx]: We should probably make sure we have all widths covered for screenshot testing purposes, even if they're not all visible on the docs site.

Comment (michaeltaranto) [reply lib/components/Column/Column.docs.tsx]: Done 👌

**Pull request merged**